runtime/queries/metricsview_aggregation.go

```diff
@@ -579,32 +584,127 @@ func (q *MetricsViewAggregation) buildMetricsAggregationSQL(mv *runtimev1.Metric
+			WHERE t2.m1 > 10
+			ORDER BY ...
+			LIMIT 100
+			OFFSET 0
+
+			This JOIN mirrors functionality of SELECT d1, d2, d3, m1 FILTER (WHERE d4 = 'Safari') FROM t WHERE... GROUP BY d1, d2, d3
+			bacause FILTER cannot be applied for arbitrary measure, ie sum(a)/1000
+		*/
+		if filterCount == 1 {
+			joinConditions := make([]string, 0, len(q.Dimensions))
+			selfJoinCols := make([]string, 0, len(q.Dimensions)+1)
+
+			selfJoinTableAlias := tempName("self_join")
+			nonNullValue := tempName("non_null")
+			for _, d := range q.Dimensions {
+				joinConditions = append(joinConditions, fmt.Sprintf("COALESCE(%[1]s.%[2]s, '%[4]s') = COALESCE(%[3]s.%[2]s, '%[4]s')", mv.Table, safeName(d.Name), selfJoinTableAlias, nonNullValue))
+				selfJoinCols = append(selfJoinCols, fmt.Sprintf("%s.%s", safeName(mv.Table), safeName(d.Name)))
+			}
+			if dialect == drivers.DialectDruid { // Apache Druid cannot order without timestamp or GROUP BY
+				selfJoinCols = append(selfJoinCols, fmt.Sprintf("ANY_VALUE(%[1]s.%[2]s) as %[3]s", selfJoinTableAlias, safeName(q.Measures[0].Name), safeName(q.Measures[0].Name)))
+			} else {
+				selfJoinCols = append(selfJoinCols, fmt.Sprintf("%[1]s.%[2]s as %[3]s", selfJoinTableAlias, safeName(q.Measures[0].Name), safeName(q.Measures[0].Name)))
+			}
 
-	return sql, args, nil
-}
+			measureExpression, measureWhereArgs, err := buildExpression(mv, q.Measures[0].Filter, nil, dialect)
+			if err != nil {
+				return "", nil, err
+			}
 
-func applyFilter(mv *runtimev1.MetricsViewSpec, expr string, filter *runtimev1.Expression, args *[]any, dialect drivers.Dialect) (string, error) {
-	if filter != nil {
-		exprClause, exprArgs, err := buildExpression(mv, filter, nil, dialect)
-		if err != nil {
-			return "", err
+			if whereClause == "" {
+				whereClause = "WHERE 1=1"
+			}
+
+			measureWhereClause := whereClause + fmt.Sprintf(" AND (%s)", measureExpression)
+			if len(sortingCriteria) > 0 {
+				jsc := make([]string, 0, len(sortingCriteria))
+				for _, v := range sortingCriteria {
+					jsc = append(jsc, fmt.Sprintf("%s.%s", safeName(mv.Table), v))
+				}
+				orderClause = "ORDER BY " + strings.Join(jsc, ", ")
+			}
+			var extraWhere string
+			var extraWhereArgs []any
+			if q.Having != nil {
+				aliases := [1]*runtimev1.MetricsViewComparisonMeasureAlias{
+					{
+						Name:  fmt.Sprintf("%s.%s", selfJoinTableAlias, q.Measures[0].Name),
+						Alias: q.Measures[0].Name,
+					},
+				}
+				extraWhere, extraWhereArgs, err = buildExpression(mv, q.Having, aliases[:], dialect)
+				if err != nil {
+					return "", nil, err
+				}
+				extraWhere = "WHERE " + extraWhere
+			}
+			druidGroupBy := ""
+			if dialect == drivers.DialectDruid {
+				druidGroupBy = groupClause
+			}
+
+			sql = fmt.Sprintf(`
+					SELECT %[1]s FROM (
+						SELECT %[10]s FROM %[2]s %[3]s %[4]s %[5]s %[6]s 
+					) %[2]s 
+					LEFT JOIN (
+						SELECT %[10]s FROM %[2]s %[3]s %[9]s %[5]s %[6]s
```

Comment:
`%[6]s` represents the having clause – does it make sense to have it in the subqueries when it will be enforced after the left join? 

At least in the first sub-query, we should be able to completely remove calculation of the measure (see other comment), so having it there seems weird.

Reply:
assuming Druid won't push down the WHERE, the subselect generates more rows for joining